### PR TITLE
Refactor(server): make ai integration more flexible

### PR DESCRIPTION
**Pull request by abrl91**
Close: https://github.com/amplication/private-issues/issues/127

## PR Details

[server]: trigger break > start conversation > create user action by type with initial step >  emit:conversation start
[gpt-gateway]: on conversation start > processing > emit: conversation completed
[server]: on conversation completed > find the user action ID and update the action step status and the user action metadata

## PR Checklist

- [x] Tests for the changes have been added
- [x] `npm test` doesn't throw any error

IMPORTANT: Please review the [CONTRIBUTING.md](https://github.com/amplication/amplication/blob/master/CODE_OF_CONDUCT.md) file for detailed contributing guidelines.

### CODE REVIEW

File: packages/amplication-prisma-db/prisma/migrations/20240119173706_remove_btm_recommendations_tables/migration.sql [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+
+-- DropForeignKey
+ALTER TABLE "BtmEntityRecommendation" DROP CONSTRAINT "BtmEntityRecommendation_originalEntityId_fkey";
+
+-- DropForeignKey
+ALTER TABLE "BtmResourceRecommendation" DROP CONSTRAINT "BtmResourceRecommendation_resourceId_fkey";
+
+-- DropTable
+DROP TABLE "BtmEntityRecommendation";
+
```
Comment (yuval-hazaz) [inline packages/amplication-prisma-db/prisma/migrations/20240119173706_remove_btm_recommendations_tables/migration.sql:19]: Let's get rid of this migration and also delete the previous one where we initially created the table(assuming it was never deployed)

File: packages/amplication-server/src/core/gpt/gpt.controller.ts [hunk 2]
```diff
@@ -0,0 +1,36 @@
+import { GptService } from "./gpt.service";
+import {
+  AiConversationComplete,
+  KAFKA_TOPICS,
+} from "@amplication/schema-registry";
+import { AmplicationLogger } from "@amplication/util/nestjs/logging";
+import { Controller, Inject } from "@nestjs/common";
+import { EventPattern, Payload } from "@nestjs/microservices";
+
+@Controller("ai-controller")
```
Comment (yuval-hazaz) [inline packages/amplication-server/src/core/gpt/gpt.controller.ts:10]: should we change this value?

File: packages/amplication-server/src/core/resource/dto/BreakServiceToMicroserviceResult.ts [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+})
+export class BreakTheMonolithRecommendationsResult {
+  @Field(() => [Microservice])
+  microservices: Microservice[];
+}
+
+@ObjectType({
+  isAbstract: true,
+})
+export class BreakServiceToMicroserviceResult {
```
Comment (yuval-hazaz) [inline packages/amplication-server/src/core/resource/dto/BreakServiceToMicroserviceResult.ts:40]:
```suggestion
export class BreakServiceToMicroservicesResult {
```
also, update the file name

File: packages/amplication-server/src/core/resource/dto/BreakServiceToMicroserviceResult.ts [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+    nullable: false,
+    description: "The status of the user action",
+  })
+  status: EnumUserActionStatus;
+
+  @Field(() => String, {
+    nullable: false,
+    description: "The original resource ID",
+  })
+  originalResourceId: string;
```
Comment (yuval-hazaz) [inline packages/amplication-server/src/core/resource/dto/BreakServiceToMicroserviceResult.ts:51]:
```suggestion
  resourceId: string;
```

File: packages/amplication-server/src/core/resource/dto/BreakServiceToMicroserviceResult.ts [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+
+  @Field(() => BreakTheMonolithRecommendationsResult, {
+    nullable: true,
+    description: "Prompt result with some data manipulation",
+  })
+  data: BreakTheMonolithRecommendationsResult;
```
Comment (yuval-hazaz) [inline packages/amplication-server/src/core/resource/dto/BreakServiceToMicroserviceResult.ts:57]:
```suggestion
  data: BreakServiceToMicroservicesData;
```

File: packages/amplication-server/src/core/resource/dto/BreakServiceToMicroserviceResult.ts [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+  functionality: string;
+
+  @Field(() => [BreakTheMonolithDataModel])
+  dataModels: BreakTheMonolithDataModel[];
+}
+
+@ObjectType({
+  isAbstract: true,
+})
+export class BreakTheMonolithRecommendationsResult {
```
Comment (yuval-hazaz) [inline packages/amplication-server/src/core/resource/dto/BreakServiceToMicroserviceResult.ts:32]:
```suggestion
export class BreakServiceToMicroservicesData {
```

File: packages/amplication-server/src/core/resource/dto/BreakServiceToMicroserviceResult.ts [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+  originalEntityId: string;
+
+  @Field()
+  name: string;
+}
+
+@ObjectType({
+  isAbstract: true,
+})
+class Microservice {
```
Comment (yuval-hazaz) [inline packages/amplication-server/src/core/resource/dto/BreakServiceToMicroserviceResult.ts:18]:
```suggestion
class BreakServiceToMicroservicesItem {
```

File: packages/amplication-server/src/core/resource/dto/BreakServiceToMicroserviceResult.ts [hunk 8]
```diff
@@ -0,0 +1,58 @@
+import { Field, ObjectType } from "@nestjs/graphql";
+import { EnumUserActionStatus } from "../../userAction/types";
+
+@ObjectType({
+  isAbstract: true,
+})
+class BreakTheMonolithDataModel {
```
Comment (yuval-hazaz) [inline packages/amplication-server/src/core/resource/dto/BreakServiceToMicroserviceResult.ts:7]:
```suggestion
class BreakServiceToMicroservicesItemEntities {
```

File: packages/amplication-server/src/core/resource/errors/AiBadFormatResponseError.ts [hunk 9]
```diff
@@ -1,10 +1,7 @@
 export class AiBadFormatResponseError extends Error {
```
Comment (yuval-hazaz) [inline packages/amplication-server/src/core/resource/errors/AiBadFormatResponseError.ts:1]: Gpt?

File: packages/amplication-server/src/core/resource/resourceBtm.resolver.ts [hunk 10]
```diff
@@ -1,50 +1,49 @@
-import { UserEntity } from "../../decorators/user.decorator";
-import { GqlResolverExceptionsFilter } from "../../filters/GqlResolverExceptions.filter";
-import { GqlAuthGuard } from "../../guards/gql-auth.guard";
-import { Resource, User } from "../../models";
-import { AiService } from "./ai.service";
 import { UseFilters, UseGuards } from "@nestjs/common";
 import { Args, Mutation, Resolver, Query } from "@nestjs/graphql";
+import { Resource, User } from "../../models";
```
Comment (yuval-hazaz) [inline packages/amplication-server/src/core/resource/resourceBtm.resolver.ts:3]: reviewed until here---------

**Review by yuval-hazaz — CHANGES_REQUESTED** [review]

File: packages/amplication-prisma-db/prisma/migrations/20240119173706_remove_btm_recommendations_tables/migration.sql (hunk 1 above)
Comment (abrl91) [reply packages/amplication-prisma-db/prisma/migrations/20240119173706_remove_btm_recommendations_tables/migration.sql:19]: It was deployed to sandbox

File: packages/amplication-server/src/core/resource/resourceBtm.service.ts [hunk 11]
```diff
@@ -0,0 +1,261 @@
...
+    OptionSet: "enum",
+    Boolean: "bool",
+    GeographicLocation: "string",
+    Id: "int",
+    CreatedAt: "datetime",
+    UpdatedAt: "datetime",
+    Roles: "string",
+    Username: "string",
+    Password: "string",
+    Json: "string",
```
Comment (yuval-hazaz) [inline packages/amplication-server/src/core/resource/resourceBtm.service.ts:43]: do we need this? please check the results against GPT with or without it and if this is not needed let's get rid of it

File: packages/amplication-server/src/core/resource/resourceBtm.service.ts [hunk 12]
```diff
@@ -0,0 +1,261 @@
...
+
+    const userAction = await this.gptService.startConversation(
+      ConversationTypeKey.BreakTheMonolith,
+      conversationParams,
+      userId,
+      resourceId
+    );
+
+    return userAction;
+  }
```
Comment (yuval-hazaz) [inline packages/amplication-server/src/core/resource/resourceBtm.service.ts:78]: Perfect

File: packages/amplication-server/src/core/resource/resourceBtm.service.ts [hunk 13]
```diff
@@ -0,0 +1,261 @@
...
+
+  duplicatedEntities(entities: string[]): Set<string> {
+    return new Set(
+      entities.filter((entity, index) => {
+        return entities.indexOf(entity) !== index;
+      })
+    );
+  }
+
+  async getResourceDataForBtm(resourceId: string): Promise<ResourcePartial> {
```
Comment (yuval-hazaz) [inline packages/amplication-server/src/core/resource/resourceBtm.service.ts:222]:
```suggestion
  async getResourceDataForBtm(resourceId: string): Promise<ResourceDataForBtm> {
```

File: packages/amplication-server/src/core/resource/resourceBtm.types.ts [hunk 14]
```diff
@@ -15,3 +15,23 @@ export interface EntityPartial
 export interface ResourcePartial extends Pick<Resource, "id" | "name"> {
```
Comment (yuval-hazaz) [inline packages/amplication-server/src/core/resource/resourceBtm.types.ts:15]:
```suggestion
export interface ResourceDataForBtm extends Pick<Resource, "id" | "name"> {
```

File: packages/amplication-server/src/core/resource/resourceBtm.service.ts [hunk 15]
```diff
@@ -0,0 +1,261 @@
...
+            };
+          }),
+        };
+      }),
+    };
+
+    return JSON.stringify(prompt);
+  }
+
+  async preparePromptResultToBtmRecommendation(
```
Comment (yuval-hazaz) [inline packages/amplication-server/src/core/resource/resourceBtm.service.ts:139]: This is not PROMPT - this is the results
```suggestion
  async prepareBtmRecommendations(
```

File: packages/amplication-server/src/core/resource/resourceBtm.service.ts [hunk 16]
```diff
@@ -0,0 +1,261 @@
...
+                originalEntityId: entityNameIdMap[dataModelName]?.id,
+              };
+            }),
+        })),
+    };
+  }
+
+  mapToBreakTheMonolithPromptOutput(
+    promptResult: string
+  ): BreakTheMonolithPromptOutput {
```
Comment (yuval-hazaz) [inline packages/amplication-server/src/core/resource/resourceBtm.service.ts:198]: This is not promptOutopt...

maybe just "BreakTheMonolithOutput"

File: packages/gpt-gateway/src/conversationType/conversationType.resolver.ts [hunk 17]
```diff
@@ -35,8 +35,7 @@ export class ConversationTypeResolver extends ConversationTypeResolverBase {
     const conversation = await this.service.startConversionSync(args);
     return {
       success: conversation.success,
-      actionId: args.actionId,
-      requestUniqueId: conversation.requestUniqueId,
+      userActionId: args.userActionId,
```
Comment (yuval-hazaz) [inline packages/gpt-gateway/src/conversationType/conversationType.resolver.ts:38]: The GPT gateway should not know anything about user action.
it expose the requestUniqueId to allow the client to use any type of ID (can be userActionId can can be anything else)
```suggestion
      requestUniqueId: conversation.requestUniqueId,
```

File: packages/gpt-gateway/src/conversationType/conversationType.service.ts [hunk 18]
```diff
@@ -24,10 +24,9 @@ export class ConversationTypeService extends ConversationTypeServiceBase {
 
   async startConversionSync(message: AiConversationStart.Value): Promise<{
     success: boolean;
-    requestUniqueId: string;
     result: string;
   }> {
-    const { messageTypeKey, params, requestUniqueId } = message;
```
Comment (yuval-hazaz) [inline packages/gpt-gateway/src/conversationType/conversationType.service.ts:30]: why did you remove this from here?

**Review by yuval-hazaz — CHANGES_REQUESTED** [review]
mainly look on the changes on the GTP service - I don't think we should do any changes in the GTP gateway service for this PR

File: packages/amplication-server/src/core/resource/dto/BreakServiceToMicroserviceResult.ts (hunk 4 above)
Comment (abrl91) [reply packages/amplication-server/src/core/resource/dto/BreakServiceToMicroserviceResult.ts:51]: I think `originalResourceId` provides more context and it's consistent with `originalEntityId` that we use in `BreakTheMonolithDataModel`

File: packages/amplication-server/src/core/resource/resourceBtm.service.ts (hunk 16 above)
Comment (abrl91) [reply packages/amplication-server/src/core/resource/resourceBtm.service.ts:198]: it sounds like a variable name and not a function that does something

Comment (yuval-hazaz) [reply packages/amplication-server/src/core/resource/resourceBtm.service.ts:198]: I was talking about the return type, but same about the function name "mapToBreakTheMonolithOutput" (the prompt is the text we send to GPT, it is not the results that returned from it)

File: packages/amplication-server/src/core/resource/resourceBtm.service.ts [hunk 19]
```diff
@@ -0,0 +1,261 @@
...
+          functionality: microservice.functionality,
+          dataModels: microservice.dataModels,
+        })),
+      };
+    } catch (error) {
+      throw new GptBadFormatResponseError(JSON.stringify(promptResult), error);
+    }
+  }
+
+  duplicatedEntities(entities: string[]): Set<string> {
```
Comment (yuval-hazaz) [inline packages/amplication-server/src/core/resource/resourceBtm.service.ts:214]: Should this be named "findDuplicatedEntities"?

File: packages/amplication-server/src/core/resource/resourceBtm.service.ts (hunk 16 above)
Comment (abrl91) [reply packages/amplication-server/src/core/resource/resourceBtm.service.ts:198]: That's why there is "output" there...but I can change it

File: packages/amplication-prisma-db/prisma/migrations/20240119173706_remove_btm_recommendations_tables/migration.sql (hunk 1 above)
Comment (mulygottlieb) [reply packages/amplication-prisma-db/prisma/migrations/20240119173706_remove_btm_recommendations_tables/migration.sql:19]: Since this was deployed to sandbox, the overhead to change it manually is not worth it IMO. Let's stay with the migrations as-is.

File: packages/amplication-server/src/core/resource/resourceBtm.service.ts (hunk 11 above)
Comment (mulygottlieb) [reply packages/amplication-server/src/core/resource/resourceBtm.service.ts:43]: @yuval-hazaz, Dani played a lot with the prompt (coordinated with Paz), to get the results we have now. While I can't say confidently that this will change the results, I think Dani will have more insight into this, and if we test on a single schema, we might not know (Dani tested more extensively until he reached\ this prompt).
Therefore, I suggest leaving it as-is for now (seeing also that the cost of removing this later is 2 minutes of work), and waiting to consult with Dani when he returns. We can open a ticket to remind us to look into this again and evaluate if we need this mapping or not.

**Review by yuval-hazaz — APPROVED** [review]
